### DISCUSSION TITLE
Modpoll su HA

### DISCUSSION THREAD

**Discussion by paolomorabito** [indomus/forum#1560] [HUB: Home Assistant]
Buongiorno a tutti,
avendo avuto sempre diversi problemi con il modbus di Hassio (a volte risolti usando le librerie modbus di Nodered), mi chiedevo se fosse possibile eseguire la versione linux di modpoll (https://www.modbusdriver.com/modpoll.html) su HA, acquisendo in qualche modo i valori di output. Chiedo questo perché Modpoll ha sempre funzionato in tutte le situazioni, al contrario di HA. Se sì, come è possibile caricarlo/installarlo su HA?

Grazie

**Comment by paolomorabito** [comment]
Ho provato a caricare al versione x86_64-linux-gnu di modpoll su /config/temp di HA, ho dato i permessi di esecuzione e.... FUNZIONA!
Nella cartella /config/temp eseguo il comando: 

`./modpoll -t 4:float -f -r 45166 -o 5 -i -c 4 -1 192.168.1.120`

e ho i risultati corretti!

Qualcuno potrebbe dirmi come fare a creare uno script da HA che mi esegue ogni 60 secondi quel comando?

**Reply by snakuzzo** [reply]
[command line sensor](https://www.home-assistant.io/integrations/sensor.command_line/)

**Reply by paolomorabito** [reply]
Ma per richiamare quella riga, come faccio? Devo impostare una path da qualche parte. Non sono esperto di linux, ma ovviamente, se provo a usare il comando `./modpoll -t 4:float -f -r 45166 -o 5 -i -c 4 -1 192.168.1.120`  su:

`sensor:
  - platform: command_line
    name: mySensor
    command: "./modpoll -t 4:float -f -r 45166 -o 5 -i -c 4 -1 192.168.1.120"`

Non funziona perchè non c'è la path di modpoll

**Reply by snakuzzo** [reply]
Dagli il path assoluto

```
/config/temp modpoll ...
```

**Reply by paolomorabito** [reply]
Già provato con `/config/temp modpoll,` non va. Mi dice `permission denied`
Sicuramente deve essere scritto in altra forma

**Reply by snakuzzo** [reply]
Ma NON così...guarda bene cosa ti ho scritto

**Reply by snakuzzo** [reply]
Scusa...m'ero perso unp `/` :)

```
/config/temp/modpoll ...
```

**Reply by paolomorabito** [reply]
Sì sì, già provato con `/config/temp/modpoll` Allora:
1. Se lo eseguo sul terminale SSH, funziona
2. Se lo eseguo dalla shell presente su FILE EDITOR, mi esce errore

**Reply by paolomorabito** [reply]
Niente. Ho provato a copiare modpoll anche su /etc  e su /bin ... ma niente. Lo eseguo dal terminale, ma non come comando della shell presente su file editor

**Reply by snakuzzo** [reply]
Scusa eh ma che installazione hai?
Stai facendo prove a caso...e così non va bene

**Reply by paolomorabito** [reply]
Purtroppo non ho molta dimestichezza con linux. 
Ho Hassio su una macchina virtuale VMWare. Hassos 7.4 e Core/supervisor 2022.3.3

Giusto per precisione, dopo aver copiato modpoll su /bin ho questa situazione.

Se da HA apro la finestra del terminale ed eseguo `modpoll -t 4:float -f -r 45166 -o 5 -i -c 4 -1 192.168.1.120` tutto funziona:
`modpoll -t 4:float -f -r 45166 -o 5 -i -c 4 -1 192.168.1.120

modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP, FC3
Slave configuration...: address = 1, start reference = 45166, count = 4
Communication.........: 192.168.1.120, port 502, t/o 5.00 s, poll rate 1000 ms
Data type.............: 32-bit float, output (holding) register table
Word swapping.........: Slave configured as big-endian word and float machine

-- Polling slave...
[45166]: 3511.693115
[45168]: 0.000000
[45170]: 746.195984
[45172]: 90.900002`

Però se inserisco quella riga su un sensore mi dà errore:

`Command failed: modpoll -t 4:float -f -r 45166 -o 5 -i -c 4 -1 192.168.1.120
18:19:31 – (ERRORE) command_line - il messaggio si è verificato per la prima volta alle 18:06:08 e compare 13 volte`

**Reply by snakuzzo** [reply]
Innanzi tutto il file `modpoll` lo puoi posizionare dove ti pare (salvo particolari esigenze legate a path relativi, etc).
Premetto che non conosco questo tool, ma da quello che vedo, quando lanci il comando, resta lì appeso a monitorare...o sbaglio?
Se così fosse, non va assolutamente bene...devi fare in modo che il comando che chiami termini e restituisca in output un "qualcosa" che nel caso dovresti adeguare in base al valore che vuoi visualizzare con il sensore creato su HA.

**Reply by paolomorabito** [reply]
> Innanzi tutto il file `modpoll` lo puoi posizionare dove ti pare (salvo particolari esigenze legate a path relativi, etc). Premetto che non conosco questo tool, ma da quello che vedo, quando lanci il comando, resta lì appeso a monitorare...o sbaglio? Se così fosse, non va assolutamente bene...devi fare in modo che il comando che chiami termini e restituisca in output un "qualcosa" che nel caso dovresti adeguare in base al valore che vuoi visualizzare con il sensore creato su HA.

No, Modpoll viene eseguito, restituisce un output e termina. Non rimane attivo. Devo richiamarlo ogni volta che mi serve aggiornare i sensori.

Però,  non vorrei sbagliarmi, ma da qualche parte avevo letto che non è possibile far partire certi eseguibili dalla shell dell'editor, e quindi all'interno di un sensore command_line

Credo ci voglia qualche altra soluzione

**Comment by paolomorabito** [comment]
Ho risolto!

Basta copiare l'eseguibile `modpoll` in una cartella qualsiasi. Io l'ho caricato su `/config/packages`
Fatto questo, basta creare un sensore nel modo seguente:

`  - platform: command_line
    name: mioSensore
    command: /config/packages/modpoll -t 4:float -f -r 45166 -o 5 -i -c 1 -1 192.168.1.120 | grep '^\['  
    value_template: "{{ value.split()[1] }}"
    scan_interval: 30
    `


Il mio errore era quello di cercare di prendere tutto l'output del comando. Questo non fa funzionare correttamente il sensore. Pertanto con `| grep '^\['  ` filtro direttamente il registro che mi interessa leggere. Invece con `value_template: "{{ value.split()[1] }}"` prendo il valore.

**Reply by snakuzzo** [reply]
ed infatti è quello che intendevo nel mio ultimo messaggio :)
Bene così ;)
